--- pascal.y
+++ pascal.y
@@ -246,7 +246,7 @@
 
 constants: 
          | PCONST constdefinitions
-         | PCONST constdefinitions conststringdefinition
+         | PCONST constdefinitions conststringdefinition constdefinitions
          ; 
 
 constdefinitions:  constdefinition
@@ -548,9 +548,9 @@
 
 
 proc_stmt: PPROCID POPEN args PCLOSE {CHGTAG($1,PCALLID); $$=$1; 
-           UP($2,$1); pstring_args($1,$3); }
+           UP($2,$1); }
          | PCALLID POPEN args PCLOSE  
-           {$$=$1; UP($2,$1); pstring_args($1,$3); }
+           {$$=$1; UP($2,$1); }
          | PPROCID  {CHGTAG($1,PCALLID); $$=$1; }
          | PCALLID  { $$=$1; }
          ;
@@ -615,9 +615,7 @@
 empty_stmt: {$$=join(CEMPTY,NULL,NULL,0); } ;
 
 assign_stmt: variable PASSIGN expression {$$=$2; }
-           | variable PASSIGN STRING {$$=$2; pstring_assign($1,$3); }
-           | variable PASSIGN POPEN STRING PCLOSE
-             {$$=$2;  pstring_assign($1,$4); }
+           | variable PASSIGN STRING {$$=$2; }
            ;
 
 return_stmt: PFUNCID PASSIGN expression  { $$=$1; }
@@ -665,6 +663,7 @@
            ;
 
 case_label: PINTEGER { winsert_after($1->previous,CCASE,"case "); }
+          | PMINUS PINTEGER { winsert_after($1->previous,CCASE,"case "); }
           | NMACRO { winsert_after($1->previous,CCASE,"case "); }
           | PINTEGER PPLUS NMACRO  
             {winsert_after($1->previous,CCASE,"case "); }
@@ -692,6 +691,11 @@
                 DBG(dbgfor,"for variable %s, limit variable in line %d\n",
                 SYM($2)->name,$2->lineno);
                 $$=join(PFOR,$8,NULL,0);LNK($1,$5);LNK($5,$7); } 
+        | PFOR PID PASSIGN expression PDOWNTO varlimit PDO statement
+              { mark_for_variable($2,$1->lineno,0,VAR_LOOP);
+                DBG(dbgfor,"for variable %s, limit variable in line %d\n",
+                SYM($2)->name,$2->lineno);
+                $$=join(PFOR,$8,NULL,0);LNK($1,$5);LNK($5,$7); }
         | PFOR PID PASSIGN expression PTO iconst PDO statement 
 	      { mark_for_variable($2,$1->lineno,$6->value,TO_LOOP); 
                 DBG(dbgfor,"for variable %s, limit up in line %d\n",
@@ -732,6 +736,7 @@
 expression: simple_expr {$$=$1; }
           | simple_expr relop simple_expr {$$=$3; }
           | simple_expr PEQ STRING {$$=$3; }
+          | simple_expr PNOTEQ STRING {$$=$3; }
           ;
 
 relop: PEQ | PNOTEQ | PLESS | PLESSEQ | PGREATER | PGREATEREQ;
